cdap-security/src/main/java/io/cdap/cdap/security/encryption/guice/LazyDelegateAeadCipher.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.security.encryption.guice;
+
+import io.cdap.cdap.security.encryption.AeadCipher;
+import io.cdap.cdap.security.spi.encryption.AeadCipherContext;
+import io.cdap.cdap.security.spi.encryption.AeadCipherCryptor;
+import io.cdap.cdap.security.spi.encryption.CipherException;
+import io.cdap.cdap.security.spi.encryption.CipherInitializationException;
+import java.util.Map;
+
+/**
+ * Default implementation of {@link AeadCipher} which delegates calls to an instance of the SPI
+ * {@link AeadCipherCryptor} with lazy initialization.
+ */
+public class LazyDelegateAeadCipher implements AeadCipher {
+
+  private final AeadCipherCryptor aeadCipherCryptor;
+  private final Map<String, String> properties;
+  private Map<String, String> secureProperties;
+  private volatile boolean initialized;
+
+  protected LazyDelegateAeadCipher(AeadCipherCryptor aeadCipherCryptor,
+      Map<String, String> properties,
+      Map<String, String> secureProperties) {
+    this.aeadCipherCryptor = aeadCipherCryptor;
+    this.properties = properties;
+    this.secureProperties = secureProperties;
```

Comment:
Make an immutable copy of incoming maps.

Reply:
Done